src/lib.rs

```diff
@@ -1,41 +1,308 @@
+    /// * `verifier` - MUST differ from the claim's subject or issuer.
+    ///
+    /// # Arguments:
+    /// * `verifier` - the claim verifier's public key.
+    /// * `verifier_signature` - the corresponding signature with the `verifier` public key.
+    /// * `piece_id` - claim's id.
+    /// * `subject` - subject's public key.
+    fn verify_claim(
+        &mut self,
+        verifier: PublicKey,
+        verifier_signature: Signature,
+        subject: PublicKey,
+        piece_id: PieceId,
+    ) {
+        if verifier == ZERO_KEY || subject == ZERO_KEY || verifier_signature == ZERO_SIGNATURE {
+            panic!("IDENTITY: Can not use a zero public key");
+        }
+        let piece = self
+            .user_claims
+            .get(&subject)
+            .expect("The user has no claims")
+            .get(&piece_id)
+            .expect("The user has not such claim with the provided piece_id");
+        if piece.issuer == verifier || piece.subject == verifier {
+            panic!("IDENTITY: You can not verify this claim");
+        }
+        self.user_claims
+            .entry(subject)
+            .or_default()
+            .entry(piece_id)
+            .and_modify(|claim| {
+                claim.verifiers.insert(verifier, verifier_signature);
+            });
+        msg::reply(
+            IdentityEvent::VerifiedClaim {
+                verifier,
+                subject,
+                piece_id,
+            },
+            0,
+        )
+        .expect("IDENTITY: Error during replying with IdentityEvent::VerifiedClaim");
+    }
 
-        let res = program.send_bytes(42, "Hello");
-        let log = Log::builder().source(1).dest(42).payload_bytes("World");
-        assert!(res.contains(&log));
+    /// Check the claim's internal data.
+    ///
+    /// # Requirements:
+    /// * all the public keys and signatures MUST be non-zero.
+    /// * `verifier` - MUST differ from the claim's subject or issuer.
+    ///
+    /// # Arguments:
+    /// * `piece_id` - claim's id.
+    /// * `subject` - subject's public key.
+    /// * `hash` - the hash to check against.
+    fn check_claim(&mut self, subject: PublicKey, piece_id: PieceId, hash: [u8; 32]) {
+        // TODO!: Rewrite in rust
+        let mut status = false;
+        if self
+            .user_claims
+            .get(&subject)
+            .expect("The user has no claims")
+            .get(&piece_id)
+            .expect("The user has not such claim with the provided piece_id")
+            .data
+            .hashed_info
+            .contains(&hash)
+        {
+            status = true;
+        }
+        msg::reply(
+            IdentityEvent::CheckedClaim {
+                subject,
+                piece_id,
+                status,
+            },
+            0,
+        )
+        .expect("IDENTITY: Error during replying with IdentityEvent::CheckedClaim");
     }
 }
+
+#[no_mangle]
+pub unsafe extern "C" fn init() {
```

Comment:
Why `pub`? `unsafe` is not necessary too.
```suggestion
extern "C" fn init() {
```

Reply:
I don't see it is resolved.

Reply:
my bad, sorry.